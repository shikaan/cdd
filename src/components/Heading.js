import React, {Component, Fragment} from 'react'
import PropTypes from 'prop-types';
import styled from 'styled-components'

const SubHeading = styled.small`
  display: block;
`

const H1 = styled.h1(({theme}) => `
  font-family: ${theme.typography.primaryFont.fontFamily};
  font-size: ${theme.typography.baseFontSize.multiply(1.5)};
  letter-spacing: 0;
  font-weight: 900;
  line-height: ${theme.typography.baseLineHeight};
  color: ${theme.color.black};
  
  & > small {
    font-family: ${theme.typography.secondaryFont.fontFamily};
    font-size: ${theme.typography.baseFontSize}
    letter-spacing: ${theme.typography.baseFontSize.multiply(0.01)};
    font-weight: 400;
    line-height: ${theme.typography.baseLineHeight};
    color: ${theme.color.black}  
  }
`)

const H2 = styled.h2(({theme}) => `
  font-family: ${theme.typography.primaryFont.fontFamily};
  font-size: ${theme.typography.baseFontSize.multiply(1.25)};
  letter-spacing: 0;
  font-weight: 900;
  line-height: ${theme.typography.baseLineHeight};
  color: ${theme.color.black};
  
  & > small {
    font-family: ${theme.typography.secondaryFont.fontFamily};
    font-size: ${theme.typography.baseFontSize.multiply(0.875)}
    letter-spacing: ${theme.typography.baseFontSize.multiply(0.00625)};
    font-weight: 400;
    line-height: ${theme.typography.baseLineHeight};
    color: ${theme.color.black}  
  }
`)

const headingMap = {
  1: H1,
  2: H2,
}


class Heading extends Component {
  render() {
    const {
      children,
      className,
      level = 1,
      sub
    } = this.props

    const Heading = headingMap[level]

    return (
      <Fragment>
        <Heading className={className}>
          {children}
        </Heading>
        {sub && <SubHeading className={className}>{sub}</SubHeading>}
      </Fragment>
    )
  }
}

Heading.propTypes = {
  level: PropTypes.oneOf([1, 2]),
  sub: PropTypes.string
};

export default Heading;
