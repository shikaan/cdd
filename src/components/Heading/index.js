import React from 'react'
import styled from 'styled-components'

import { size } from '~style/variables'
import { baseLineHeight, primaryFont, secondaryFont } from '~style/typography'

const H1 = styled.h1({
  fontFamily: primaryFont.fontFamily,
  lineHeight: baseLineHeight,
  fontWeight: 900,
  letterSpacing: 0,
  fontSize: size(3)
})

const H2 = styled.h2({
  fontFamily: secondaryFont.fontFamily,
  lineHeight: baseLineHeight,
  fontWeight: 400,
  letterSpacing: '0.15px',
  fontSize: size(2)
})

const headingMap = {
  1: H1,
  2: H2
}

export default ({ level = 1, children }) => {
  return React.createElement(headingMap[level], { children })
}
