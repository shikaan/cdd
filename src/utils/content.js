/**
 * @param {Array<{reference: string}>} microcopy
 * @param {string} key
 * @return {string}
 */
export const getMicrocopy = (microcopy, key) => {
  const element = microcopy?.find(i => i.reference === key) ?? {};

  return element.value ?? "";
};

/**
 * @param {Array<{reference: string}>} sections
 * @param {string} key
 * @return {string}
 */
export const getSection = (sections, key) => {
  return sections?.find(i => i.reference === key) ?? {reference: key};
};