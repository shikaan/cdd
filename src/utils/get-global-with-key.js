export const getGlobalWithKey = (key = "location") => {
  return typeof window !== "undefined" ? window : {[key]: {}};
};
