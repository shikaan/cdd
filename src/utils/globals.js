export const getGlobalWithKey = (key = "location") => {
  return typeof window !== "undefined" ? window : {[key]: {}};
};

export const getGlobal = () => { 
  // eslint-disable-next-line no-restricted-globals
  if (typeof self !== "undefined") { return self }
  if (typeof window !== "undefined") { return window } 
  if (typeof global !== "undefined") { return global } 
  throw new Error("unable to locate global object"); 
}; 
