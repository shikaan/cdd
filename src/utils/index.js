export {isLastIndex} from "./is-last-index";
