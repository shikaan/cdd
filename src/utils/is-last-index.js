export const isLastIndex = (list, index) => {
  if (!Array.isArray(list)) {
    return false;
  }

  return index === list.length - 1;
};
