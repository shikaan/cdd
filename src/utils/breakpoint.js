import {numberBreakpoint} from "~theme/variables";
import {getGlobalWithKey} from "~/utils/get-global-with-key";

export const getViewportWidth = () => getGlobalWithKey("innerWidth") ?? getGlobalWithKey("document").documentElement.clientWidth;

export const isMobile = () => {
  const viewportWidth = getViewportWidth();
  return viewportWidth < numberBreakpoint.sm;
};

export const isTablet = () => {
  const viewportWidth = getViewportWidth();
  return numberBreakpoint.sm <= viewportWidth && viewportWidth < numberBreakpoint.md;
};

export const isDesktop = () => {
  const viewportWidth = getViewportWidth();
  return numberBreakpoint.md <= viewportWidth;
};