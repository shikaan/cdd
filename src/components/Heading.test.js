import React from 'react'
import {shallow} from 'enzyme';
import 'jest-styled-components';

import themeMock from '~theme'

import Heading from "./Heading.js";
import {getTagNameByHTMLString, shallowWithTheme} from "/test/utils";
import {CONTEXT} from "./Heading";

describe('Heading', () => {
  it('renders sub if provided', () => {
    const wrapper = shallow(<Heading theme={themeMock} sub={'sub'}/>)

    expect(wrapper).toMatchSnapshot()
  })

  it('does not render sub if not provided', () => {
    const wrapper = shallow(<Heading theme={themeMock}/>)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders h$N where $N is the level provided', () => {
    for (let i = 1; i <= 3; i++) {
      const wrapper = shallowWithTheme(<Heading level={i}/>, themeMock)

      expect(getTagNameByHTMLString(wrapper.html())).toEqual(`h${i}`)
    }
  })

  it('renders correctly with CONTEXT.DISPLAY', () => {
    const wrapper = shallowWithTheme(<Heading context={CONTEXT.DISPLAY} />, themeMock)

    expect(wrapper).toMatchSnapshot()
  })
})
