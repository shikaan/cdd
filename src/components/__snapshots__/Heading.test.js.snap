// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading does not render sub if not provided 1`] = `
<Fragment>
  <styled.h1 />
</Fragment>
`;

exports[`Heading renders correctly with CONTEXT.DISPLAY 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Heading renders sub if provided 1`] = `
<Fragment>
  <styled.h1 />
  <styled.small>
    sub
  </styled.small>
</Fragment>
`;
